### PR TITLE
Bazel-ify Cortex

### PR DESCRIPTION
**Pull request by tomwilkie**
Adds all the required bits to build cortex with bazel.  Incremental rebuilds of a single small change take about 20s with bazel:

```
$ git diff
diff --git a/ring/http.go b/ring/http.go
index dcf256f..242eace 100644
--- a/ring/http.go
+++ b/ring/http.go
@@ -20,7 +20,7 @@ const tpl = `
        </head>
        <body>
                <h1>Cortex Ring Status</h1>
-               <p>Current time: {{ .Now }}</p>
+               <p>Current time:  {{ .Now }}</p>
                <p>{{ .Message }}</p>
                <form action="" method="POST">
                        <table width="100%" border="1">
$ make blaze
bazel build //cmd/...
INFO: Found 14 targets...
INFO: Elapsed time: 20.802s, Critical Path: 20.30s
```

vs our existing build which just takes so long I can't be bothered to wait.

I think this will get quicker once I've pruned the vendor dir.  Also need to investigate build linux binaries and docker containers.

### CODE REVIEW

**Comment by juliusv** [comment]
Does Bazel help create the `BUILD` files for you, or does that require manual maintenance?

**Comment by juliusv** [comment]
Btw., doing the same incremental build as indicated by your diff just takes 23s for me as well (full `make` run). I wonder what's up there for you...

**Comment by tomwilkie** [comment]
Yeah I think my VM was borked.   Also, you're on linux right?  And you're docker is not in a VM?

Going to rebase as I've removed most of vendor/, so hopefully this will be quicker.

**Comment by tomwilkie** [comment]
> Does Bazel help create the BUILD files for you, or does that require manual maintenance?

Automatically.  I think we might even be able to not commit them...  I added a makefile target to generate them.

**Comment by juliusv** [comment]
> Automatically. I think we might even be able to not commit them... I added a makefile target to generate them.

Great.

> Also, you're on linux right? And you're docker is not in a VM?

Yep and yep.

**Comment by tomwilkie** [comment]
This is now ready to go.  I had to check in the generated BUILD files as they were wrong in two instances - on dealing with the the `_thrift` generated files and the other dealing with a `.h` file in the `golang/s/crypto` package.

I also had to move all non-main packages into `pkg/` - this is something I've wanted to do for a while TBH, and was nessecitated by the desire to run tests using bazel.  Before this change, `bazel test //...` would try and run tests for the stuff in vendor; now we can just do `bazel test //pkg/...` to run all tests.

**Comment by tomwilkie** [comment]
NB the old build system is still there as it can still cross-build docker images on my Mac.  Right now the bazel build isn't super useful, except its quicker to see if you've broken anything with it.

File: Makefile [hunk 1]
```diff
@@ -114,4 +114,11 @@ clean:
 	rm -rf $(UPTODATE_FILES) $(EXES) $(PROTO_GOS)
 	go clean ./...
 
+update-gazelle:
```
Comment (jml) [inline Makefile]: Maybe add a comment about why I might want to run this target?

File: Makefile [hunk 2]
```diff
@@ -114,4 +114,11 @@ clean:
...
+	gazelle -go_prefix github.com/weaveworks/cortex -external vendored
 
+bazel:
+	bazel build //cmd/...
+
+bazel-test:
```
Comment (jml) [inline Makefile]: So we're not running bazel in CI. When would a Cortex developer want to run Bazel? How does this PR actually help?

I think it's worth putting some notes in the README about it.

**Comment by tomwilkie** [comment]
> How were they generated? 

The `update-gazelle` makefile target.

> Will they need to be regenerated at any point in the future? 

Yes; they will need to be maintained moving forward.  New files and dependancies will need to be added to them.  Unless gazelle fix the outstanding issues (I'll file upstream tickets for them) then I doubt we'd actually every regenerate them from scratch.

> How will the person who regenerates them know to fix the things you mention?

It won't build / there will be a diff.

I'm considering adding bazel to CI, but I don't want to make the CI run any longer.  And we still need the old build system.

File: Makefile (hunk 2 above)
Comment (tomwilkie) [inline Makefile]: > When would a Cortex developer want to run Bazel? How does this PR actually help?

You may want to use bazel locally to build & test, as for incremental builds & tests it faster.

**Comment by tomwilkie** [comment]
Upstream gazelle problems: 
- https://github.com/bazelbuild/rules_go/issues/422
- https://github.com/bazelbuild/rules_go/issues/423

**Pull request merged**